sktime/annotation/hmm.py

```diff
@@ -0,0 +1,275 @@
+To read more about the algorithm, check out the wikipedia page:
+(https://en.wikipedia.org/wiki/Hidden_Markov_model
+"""
+from math import isclose
+from typing import Tuple
+
+import numpy as np
+
+from sktime.annotation.base._base import BaseSeriesAnnotator
+
+__author__ = ["miraep8"]
+__all__ = ["HMM"]
+
+
+class HMM(BaseSeriesAnnotator):
+    """Implements a simple HMM fitted with viterbi algorithm.
+
+    The HMM annotation estimator facilitates the appilcation of
+    the vertibi algorithm to fit a sequence of 'hidden state' class
+    annotations (represented by an array of integers the same size
+    as the observation) to a sequence of observations.
+
+    This is done by finding the most likely path given the emission
+    probabilities - (ie the probability that a particular observation
+    would be generated by a given hidden state), the transition prob
+    (ie the probability of transitioning from one state to another or
+    staying in the same state) and the initial probabilities - ie the
+    belief of the probability distribution of hidden states at the
+    start of the observation sequence).
+
+    Parameters
+    ----------
+    emission_funcs : a list of functions, which should be properly
+        normalized PDFs over the same space as the observed data.
+    transition_prob_mat: a nxn array of probabilities, where 'n'
+        represents the number of hidden states in the HMM.  Each
+        row should sumn to 1 in order to be properly normalized
+        (ie the j'th column in the i'th row represents the
+        probability of transitioning from state i to state j.)
+    initial_probs: optional, a 1d array of probabilities across
+        the starting states. Should match prior beliefs.  If none
+        is passed will give each state an equal initial probability.
+
+    Examples
+    --------
+    >>> from sktime.annotation.hmm import HMM
+    >>> from scipy.stats import norm
+    >>> from numpy import asarray
+    >>> # define the emission probs for our HMM model:
+    >>> centers = [3.5,-5]
+    >>> sd = [.25 for i in centers]
+    >>> emi_funcs = [(norm.pdf, {'loc': mean,
+    ...  'scale': sd[ind]}) for ind, mean in enumerate(centers)]
+    >>> test = HMM(emi_funcs, asarray([[0.25,0.75], [0.666, 0.333]]))
+    >>> # generate synthetic data (or of course use your own!)
+    >>> obs = asarray([3.7,3.2,3.4,3.6,-5.1,-5.2,-4.9])
+    >>> test = test.fit(obs)
+    >>> labels = test.predict(obs)
+    """
+
+    # plan to update to make multivariate.
+    _tags = {"univariate-only": True, "fit_is_empty": False}
+
+    def __init__(
+        self,
+        emission_funcs: list,
+        transition_prob_mat: np.ndarray,
+        initial_probs: np.ndarray = None,
+    ):
+        self.initial_probs = initial_probs
+        self.num_states = len(emission_funcs)
+        self.emission_funcs = emission_funcs
+        self.transition_prob_mat = transition_prob_mat
+        self.states = [i for i in range(self.num_states)]
```

Comment:
all attributes that are not part of the `__init__` signature should be assigned in `fit` to avoid side effects when using `set_params`